### DISCUSSION TITLE
How to query data using daos

### DISCUSSION THREAD

**Discussion by markschellhas** [pocketbase/pocketbase#490] [Q&A] [answered]
I'm exploring the documentation on usage of daos (`app.Dao()`) for database queries, using PocketBase as a framework. I'd like to make database queries in a custom route, but find the documentation on this a little unclear for querying collections.

I have some records in a `posts` Collection:

For example, I'd like to query all posts, through a **custom route**:
```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/posts",
			Handler: func(c echo.Context) error {
				
				// Show all records from posts collection:
				results, err := app.Dao().DB().
				 	NewQuery("SELECT * FROM posts").
				 	Execute()
                                  if err != nil {
					return c.JSON(http.StatusInternalServerError, err)
				}
				
				return c.JSON(http.StatusOK, results)
			},
		})
})
```

This produces the following **Response**:
```JSON
{
  "Locker": {}
}
```

What am I doing wrong? What's the intended way to query data through `app.Dao().DB()`?

Generally, I'm looking for documentation on examples on how to insert, update and query data directly from the database in Go (as there is not a Go SDK) so that I can do data processing logic a Go custom route handler. Any help to clarify would be most helpful.

**Comment by ganigeorgiev** [comment, marked as answer]
Yes, the documentation is currently lacking examples how to properly work with the exposed internals. It is in the roadmap and will be revisited after the ongoing users management refactoring in #376.

You can use `app.Dao().DB()` to execute raw queries but you need to "scan" the result in order to make use of the select. Examples could be found in https://github.com/pocketbase/dbx.

But if you want to query colleciton records, there is a better way to do it - and that is to use the internal dao.Record* helpers in https://github.com/pocketbase/pocketbase/blob/master/daos/record.go (more will be added in the near future). For example, to query all `posts` collection records, you could run something like this:

```go
collection, err := app.Dao().FindCollectionByNameOrId("posts")
if err != nil {
	return err
}

// the expression could be any db.Expression (see the dbx package for more options).
records, err := app.Dao().FindRecordsByExpr(collection, dbx.HashExp{
	"status": "public",
})
if err != nil {
	return err
}
```

**Reply by markschellhas** [reply]
@ganigeorgiev Thanks for the answer. Having looked at [https://github.com/pocketbase/dbx](https://github.com/pocketbase/dbx), but the opening connections seems to show examples with `mysql` only.

How do you suggest then using this with Pocketbase though, especially regarding the CRUD examples, would you suggest opening connections like this: `db, _ := dbx.Open("sqlite3", "./pb_data/data.db")` to insert data?

```go
type Customer struct {
	ID     int
	Name   string
	Email  string
	Status int
}

db, _ := dbx.Open("sqlite3", "./pb_data/data.db")

customer := Customer{
	Name: "example",
	Email: "test@example.com",
}
// INSERT INTO customer (name, email, status) VALUES ('example', 'test@example.com', 0)
err := db.Model(&customer).Insert()
```

I'm struggling to get basic CRUD operations on the Pocketbase database from within a custom route, very much appreciate your advice.

**Reply by ganigeorgiev** [reply]
@markschellhas You don't need to open a database. What I meant was that if you want to directly work with the db abstraction (aka, dbx), you can use `app.Dao().DB()` (the DB() method returns a dbx DB instance).

But as I mentioned, it may be better to use the record daos helpers instead of writing raw sql statements.

I completely agree that the current documentation is lacking more advanced usage examples and guides and that will be fixed in the near future, but the users and profiles refactoring in #376 currently is with highest priority because it blocks all other feature requests.

**Reply by markschellhas** [reply]
@ganigeorgiev Ok, thank you very much. I'll accept this answer and continue exploring. Appreciate you taking the time to help me along this path.
